File: server/controllers/paymentController.js
import Razorpay from 'razorpay';
import crypto from 'crypto';
import Order from '../models/Order.js';
import Book from '../models/Book.js';
import { sendOrderConfirmationEmail } from '../utils/emailService.js';

// Initialize Razorpay
const razorpay = new Razorpay({
  key_id: process.env.RAZORPAY_KEY_ID,
  key_secret: process.env.RAZORPAY_SECRET
});

// Create Razorpay order
export const createPaymentOrder = async (req, res) => {
  try {
    const { orderId } = req.body;

    // Find the order
    const order = await Order.findOne({ orderId }).populate('bookId');

    if (!order) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }

    // Create Razorpay order
    const options = {
      amount: order.totalAmount * 100, // amount in paise
      currency: 'INR',
      receipt: order.orderId,
      notes: {
        orderId: order.orderId,
        bookTitle: order.bookId.title
      }
    };

    const razorpayOrder = await razorpay.orders.create(options);

    // Update order with Razorpay order ID
    order.razorpayOrderId = razorpayOrder.id;
    await order.save();

    res.status(200).json({
      success: true,
      data: {
        razorpayOrderId: razorpayOrder.id,
        amount: razorpayOrder.amount,
        currency: razorpayOrder.currency,
        keyId: process.env.RAZORPAY_KEY_ID
      }
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Error creating payment order',
      error: error.message
    });
  }
};

// Verify payment
export const verifyPayment = async (req, res) => {
  try {
    const { razorpay_order_id, razorpay_payment_id, razorpay_signature, orderId } = req.body;

    // Verify signature
    const sign = razorpay_order_id + '|' + razorpay_payment_id;
    const expectedSign = crypto
      .createHmac('sha256', process.env.RAZORPAY_SECRET)
      .update(sign.toString())
      .digest('hex');

    if (razorpay_signature !== expectedSign) {
      return res.status(400).json({
        success: false,
        message: 'Invalid payment signature'
      });
    }

    // Find and update order
    const order = await Order.findOne({ orderId }).populate('bookId');

    if (!order) {
      return res.status(404).json({
        success: false,
        message: 'Order not found'
      });
    }

    // Update order status
    order.paymentStatus = 'Paid';
    order.paymentId = razorpay_payment_id;
    order.paymentSignature = razorpay_signature;
    
    // Set delivery date (5-7 days from now)
    const deliveryDays = Math.floor(Math.random() * 3) + 5; // Random between 5-7 days
    const deliveryDate = new Date();
    deliveryDate.setDate(deliveryDate.getDate() + deliveryDays);
    order.deliveryDate = deliveryDate;

    await order.save();

    // Reduce book stock
    const book = await Book.findById(order.bookId._id);
    if (book && book.stock > 0) {
      book.stock -= 1;
      await book.save();
    }

    // Send confirmation email
    try {
      await sendOrderConfirmationEmail(order);
    } catch (emailError) {
      console.error('Email sending failed:', emailError);
      // Don't fail the payment if email fails
    }

    res.status(200).json({
      success: true,
      message: 'Payment verified successfully',
      data: {
        orderId: order.orderId,
        paymentId: razorpay_payment_id,
        deliveryDate: order.deliveryDate,
        bookTitle: order.bookId.title,
        totalAmount: order.totalAmount
      }
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Error verifying payment',
      error: error.message
    });
  }
};

// Payment failure handler
export const paymentFailed = async (req, res) => {
  try {
    const { orderId, error } = req.body;

    const order = await Order.findOne({ orderId });

    if (order) {
      order.paymentStatus = 'Failed';
      await order.save();
    }

    res.status(200).json({
      success: true,
      message: 'Payment failure recorded'
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: 'Error recording payment failure',
      error: error.message
    });
  }
};

File: server/controllers/payment_controller.py
from fastapi import HTTPException
import razorpay
import hmac
import hashlib
from datetime import datetime, timedelta
from bson import ObjectId
import random

from database import get_database
from models.order import PaymentStatus
from utils.email_service import send_order_confirmation_email
from config import settings

# Initialize Razorpay
razorpay_client = razorpay.Client(
    auth=(settings.RAZORPAY_KEY_ID, settings.RAZORPAY_SECRET)
)

async def create_payment_order(order_id: str):
    """Create Razorpay order"""
    db = get_database()
    
    # Find order by orderId
    order = await db.orders.find_one({"orderId": order_id})
    if not order:
        raise HTTPException(status_code=404, detail="Order not found")
    
    # Get book details
    book = await db.books.find_one({"_id": order["bookId"]})
    
    # Create Razorpay order
    razorpay_order = razorpay_client.order.create({
        "amount": int(order["totalAmount"] * 100),  # Amount in paise
        "currency": "INR",
        "receipt": order["orderId"],
        "notes": {
            "orderId": order["orderId"],
            "bookTitle": book["title"] if book else ""
        }
    })
    
    # Update order with Razorpay order ID
    await db.orders.update_one(
        {"orderId": order_id},
        {
            "$set": {
                "razorpayOrderId": razorpay_order["id"],
                "updatedAt": datetime.utcnow()
            }
        }
    )
    
    return {
        "razorpayOrderId": razorpay_order["id"],
        "amount": razorpay_order["amount"],
        "currency": razorpay_order["currency"],
        "keyId": settings.RAZORPAY_KEY_ID
    }

async def verify_payment(payment_data: dict):
    """Verify Razorpay payment signature"""
    razorpay_order_id = payment_data.get("razorpay_order_id")
    razorpay_payment_id = payment_data.get("razorpay_payment_id")
    razorpay_signature = payment_data.get("razorpay_signature")
    order_id = payment_data.get("orderId")
    
    # Verify signature
    generated_signature = hmac.new(
        settings.RAZORPAY_SECRET.encode(),
        f"{razorpay_order_id}|{razorpay_payment_id}".encode(),
        hashlib.sha256
    ).hexdigest()
    
    if generated_signature != razorpay_signature:
        raise HTTPException(status_code=400, detail="Invalid payment signature")
    
    db = get_database()
    
    # Find order
    order = await db.orders.find_one({"orderId": order_id})
    if not order:
        raise HTTPException(status_code=404, detail="Order not found")
    
    # Get book details
    book = await db.books.find_one({"_id": order["bookId"]})
    
    # Calculate delivery date (5-7 days)
    delivery_days = random.randint(5, 7)
    delivery_date = datetime.utcnow() + timedelta(days=delivery_days)
    
    # Update order
    await db.orders.update_one(
        {"orderId": order_id},
        {
            "$set": {
                "paymentStatus": PaymentStatus.PAID.value,
                "paymentId": razorpay_payment_id,
                "paymentSignature": razorpay_signature,
                "deliveryDate": delivery_date,
                "updatedAt": datetime.utcnow()
            }
        }
    )
    
    # Reduce book stock
    if book and book.get("stock", 0) > 0:
        await db.books.update_one(
            {"_id": book["_id"]},
            {
                "$inc": {"stock": -1},
                "$set": {"updated_at": datetime.utcnow()}
            }
        )
    
    # Get updated order
    updated_order = await db.orders.find_one({"orderId": order_id})
    updated_order["bookId"] = book  # Add book details
    
    # Send confirmation email
    try:
        await send_order_confirmation_email(updated_order)
    except Exception as e:
        print(f"Email sending failed: {e}")
    
    return {
        "orderId": order_id,
        "paymentId": razorpay_payment_id,
        "deliveryDate": delivery_date.isoformat(),
        "bookTitle": book["title"] if book else "",
        "totalAmount": updated_order["totalAmount"]
    }

async def payment_failed(order_id: str, error: str = None):
    """Record payment failure"""
    db = get_database()
    
    result = await db.orders.update_one(
        {"orderId": order_id},
        {
            "$set": {
                "paymentStatus": PaymentStatus.FAILED.value,
                "updatedAt": datetime.utcnow()
            }
        }
    )
    
    if result.modified_count == 0:
        raise HTTPException(status_code=404, detail="Order not found")
    
    return {"message": "Payment failure recorded"}
